libraries/anvill_passes/src/BranchAnalysis.cpp

```diff
@@ -0,0 +1,432 @@
+    {"zero", ArithFlags::ZF},
+    {"overflow", ArithFlags::OF},
+    {"sign", ArithFlags::SIGN},
+    {"carry", ArithFlags::CARRY}};
+
+
+struct FlagDefinition {
+  llvm::Value *lhs;
+  llvm::Value *rhs;
+  Z3Binop binop;
+  ArithFlags flagres;
+  llvm::Type *resultTy;
+};
+
+
+class EnvironmentBuilder {
+
+ private:
+  std::optional<ComparedValues> symbols;
+  std::unordered_map<std::string, FlagDefinition> bindings;
+  uint64_t id_counter;
+
+ public:
+  static std::string getValueName(uint64_t number) {
+    return "value" + std::to_string(number);
+  }
+
+ private:
+  std::string nextID() {
+    return FlagIDPrefix + std::to_string(this->id_counter++);
+  }
+
+
+  std::optional<FlagDefinition> parseFlagDefinition(RemillFlag rf) {
+    std::optional<Z3Binop> conn = std::nullopt;
+    std::optional<llvm::Value *> lhs = std::nullopt;
+    std::optional<llvm::Value *> rhs = std::nullopt;
+    if (auto *binop = llvm::dyn_cast<llvm::BinaryOperator>(rf.over)) {
+      conn = BinopExpr::TranslateOpcodeToConnective(binop->getOpcode());
+      lhs = binop->getOperand(0);
+      rhs = binop->getOperand(1);
+    } else if (auto *binop = llvm::dyn_cast<llvm::ICmpInst>(rf.over)) {
+      conn = BinopExpr::TranslateIcmpOpToZ3(binop->getPredicate());
+      lhs = binop->getOperand(0);
+      rhs = binop->getOperand(1);
+    }
+
+    if (conn && lhs && rhs) {
+      return {{*lhs, *rhs, *conn, rf.flg, rf.over->getType()}};
+    }
+
+
+    return std::nullopt;
+  }
+
+  bool composedOfSameSymbols(FlagDefinition flagdef) {
+    return !this->symbols || ((this->symbols->first == flagdef.lhs ||
+                               this->symbols->first == flagdef.rhs) &&
+                              (this->symbols->second == flagdef.lhs ||
+                               this->symbols->second == flagdef.rhs));
+  }
+
+  bool areIntegerTypes(FlagDefinition flagdef) {
+    return flagdef.lhs->getType()->isIntegerTy() &&
+           flagdef.rhs->getType()->isIntegerTy();
+  }
+
+ public:
+  EnvironmentBuilder()
+      : symbols(std::nullopt),
+        bindings(std::unordered_map<std::string, FlagDefinition>()),
+        id_counter(0) {}
+
+  inline const static std::string FlagIDPrefix = "flag";
+
+
+  std::optional<std::unique_ptr<Expr>> addFlag(RemillFlag rf) {
+    auto flagdef = this->parseFlagDefinition(rf);
+    if (flagdef.has_value() && this->composedOfSameSymbols(*flagdef) &&
+        this->areIntegerTypes(*flagdef)) {
+      this->symbols = {std::make_pair(flagdef->lhs, flagdef->rhs)};
+      auto name = this->nextID();
+      this->bindings.insert({name, *flagdef});
+      return {AtomVariable::Create(name)};
+    }
+
+    return std::nullopt;
+  }
+
+  z3::expr create_flag_for_name(z3::context &cont, std::string name) {
+    return cont.bool_const(name.c_str());
+  }
+
+  std::pair<z3::expr, z3::expr> symbols_as_z3(z3::context &cont,
+                                              Environment &env) {
+    auto v1name = EnvironmentBuilder::getValueName(0);
```

Comment:
You don't need `UserFunctionPass::` to call a static method from within a static or instance method on the same class.